pkg/common_controller/snapshot_controller.go

```diff
@@ -0,0 +1,846 @@
+	if snapshotContent.Status != nil {
+		_, err = ctrl.updateSnapshotStatus(snapshotCopy, snapshotContent)
+
+		// Update snapshot status with creationTime
+		for i := 0; i < ctrl.createSnapshotContentRetryCount; i++ {
+			klog.V(5).Infof("syncUnreadySnapshot [%s]: trying to update snapshot status", utils.SnapshotKey(snapshotCopy))
+			_, err = ctrl.updateSnapshotStatus(snapshotCopy, snapshotContent)
+			if err == nil {
+				break
+			}
+			klog.V(4).Infof("failed to update snapshot %s status: %v", utils.SnapshotKey(snapshot), err)
+		}
+
+		if err != nil {
+			// update snapshot status failed
+			ctrl.updateSnapshotErrorStatusWithEvent(snapshotCopy, v1.EventTypeWarning, "SnapshotStatusUpdateFailed", fmt.Sprintf("Snapshot status update failed, %v", err))
+			return nil, err
+		}
+	}
+
+	klog.V(5).Infof("bindandUpdateVolumeSnapshot for snapshot completed [%#v]", snapshotCopy)
+	return snapshotCopy, nil
+}
+
+// UpdateSnapshotStatus updates snapshot status based on content status
+//func (ctrl *csiSnapshotCommonController) updateSnapshotStatus(snapshot *crdv1.VolumeSnapshot, boundContentName string, readyToUse bool, createdAt time.Time, size int64, bound bool) (*crdv1.VolumeSnapshot, error) {
+func (ctrl *csiSnapshotCommonController) updateSnapshotStatus(snapshot *crdv1.VolumeSnapshot, content *crdv1.VolumeSnapshotContent) (*crdv1.VolumeSnapshot, error) {
+	if content.Status == nil {
+		return nil, fmt.Errorf("content status for snapshot %s is nil. Nothing to update", utils.SnapshotKey(snapshot))
+	}
+
+	boundContentName := content.Name
+	var createdAt *time.Time
+	if content.Status != nil && content.Status.CreationTime != nil {
+		unixTime := time.Unix(0, *content.Status.CreationTime)
+		createdAt = &unixTime
+	}
+	var size int64
+	if content.Status != nil && content.Status.RestoreSize != nil {
+		size = *content.Status.RestoreSize
+	}
+	var readyToUse bool
+	if content.Status != nil && content.Status.ReadyToUse != nil {
+		readyToUse = *content.Status.ReadyToUse
+	}
+
+	klog.V(5).Infof("updating VolumeSnapshot [%+v] based on VolumeSnapshotContentStatus [%+v]", snapshot, content.Status)
+
+	snapshotObj, err := ctrl.clientset.SnapshotV1beta1().VolumeSnapshots(snapshot.Namespace).Get(snapshot.Name, metav1.GetOptions{})
+	if err != nil {
+		return nil, fmt.Errorf("error get snapshot %s from api server: %v", utils.SnapshotKey(snapshot), err)
+	}
+
+	var newStatus *crdv1.VolumeSnapshotStatus
+	updated := false
+	if snapshotObj.Status == nil {
+		newStatus = &crdv1.VolumeSnapshotStatus{
+			BoundVolumeSnapshotContentName: &boundContentName,
+			//CreationTime:                   &metav1.Time{Time: createdAt},
+			ReadyToUse:  &readyToUse,
+			RestoreSize: resource.NewQuantity(size, resource.BinarySI),
+		}
+		if createdAt != nil {
+			newStatus.CreationTime = &metav1.Time{Time: *createdAt}
+		}
+		updated = true
+	} else {
+		newStatus = snapshotObj.Status.DeepCopy()
+		if newStatus.BoundVolumeSnapshotContentName == nil {
```

Comment:
It might worth checking whether if BoundVolumeSnapshotContentName == boundContentName to decide whether or not to continue here? That case it should be invalid?